core/src/torii/routing.rs

```diff
@@ -161,16 +161,84 @@ async fn stream_blocks<W: WorldTrait>(
     Ok(())
 }
 
-#[iroha_futures::telemetry_future]
-async fn handle_subscription(events: EventsSender, stream: WebSocket) -> eyre::Result<()> {
-    let mut events = events.subscribe();
-    let mut consumer = Consumer::new(stream).await?;
+mod subscription {
+    //! This module contains `handle_subscription()` function and other stuff for its implementation
+
+    use super::*;
+    use crate::{event, stream};
+
+    /// Type for any error during subscription handling
+    #[derive(thiserror::Error, Debug)]
+    enum Error {
+        /// Event consuming error
+        #[error("Event consuming error: {0}")]
+        Consumer(Box<event::Error>),
+        /// Event receiving error
+        #[error("Event receiving error: {0}")]
+        Event(#[from] tokio::sync::broadcast::error::RecvError),
+        /// Error from provided websocket
+        #[error("WebSocket error: {0}")]
+        WebSocket(#[from] warp::Error),
+        /// Error, indicating that `Close` message was received
+        #[error("`Close` message received")]
+        CloseMessage,
+    }
 
-    loop {
-        let event = events.recv().await?;
+    type StreamError = stream::Error<<WebSocket as Stream<VersionedEventSubscriberMessage>>::Err>;
+
+    impl From<event::Error> for Error {
+        fn from(error: event::Error) -> Self {
+            match error {
+                event::Error::Stream(StreamError::CloseMessage) => Self::CloseMessage,
+                error => Self::Consumer(Box::new(error)),
+            }
+        }
+    }
+
+    type Result<T> = core::result::Result<T, Error>;
+
+    /// Handle subscription request
+    ///
+    /// Subscribes `stream` for `events` filtered by filter that is received through the `stream`
+    ///
+    /// There should be a [`warp::filters::ws::Message::close()`] message to end subscription
+    #[iroha_futures::telemetry_future]
+    pub async fn handle_subscription(events: EventsSender, stream: WebSocket) -> eyre::Result<()> {
+        match subscribe_forever(events, stream).await {
+            Ok(()) | Err(Error::CloseMessage) => Ok(()),
+            Err(err) => Err(err.into()),
+        }
+    }
 
-        iroha_logger::trace!(?event);
-        consumer.consume(event).await?;
+    /// Make endless `stream` subscription for `events`
+    ///
+    /// Ideally should return `Result<!>` cause it either runs forever either returns `Err` variant
+    async fn subscribe_forever(events: EventsSender, stream: WebSocket) -> Result<()> {
+        let mut events = events.subscribe();
+        let mut consumer = Consumer::new(stream).await?;
+
+        loop {
+            tokio::select! {
+                // This branch catches `Close` ans unexpected messages
+                message_opt = consumer.stream_mut().next() => {
```

Comment:
Would it be better/possible to have something like `consumer.stream_closed` which would encapsulate the logic instead of exposing `stream_mut`?

Totally up to you